docs/proposal/20240308-vpc-infra-creation.md

```diff
@@ -0,0 +1,402 @@
+
+	// COSInstance is the reference to the IBM Cloud COS Instance used for the cluster.
+	COSInstance *VPCResourceReference `json:"cosInstance,omitempty"`
+
+	// networkStatus is the status of the VPC network in its entirety resources.
+	NetworkStatus *VPCNetworkStatus `json:"networkStatus,omitempty"`
+
+	// ready is true when the provider resource is ready.
+	// +kubebuilder:default=false
+	Ready bool `json:"ready"`
+
+	// resourceGroup is the reference to the IBM Cloud VPC resource group under which the resources will be created.
+	ResourceGroup *ResourceReference `json:"resourceGroupID,omitempty"`
+
+	// Deprecated: rely on NetworkStatus instead.
+	Subnet Subnet `json:"subnet,omitempty"`
+
+	// Deprecated: rely on NetworkStatus instead.
+	VPC VPC `json:"vpc,omitempty"`
+
+	// Deprecated: rely on ControlPlaneEndpoint
+	VPCEndpoint VPCEndpoint `json:"vpcEndpoint,omitempty"`
+}
+
+// VPCNetworkStatus provides details on the status of VPC network resources.
+type VPCNetworkStatus struct {
+	// computeSubnets references the VPC Subnets for the cluster's Data Plane.
+	// +optional
+	ComputeSubnets []*VPCResourceReference `json:"computeSubnets,omitempty"`
+
+	// controlPlaneSubnets references the VPC Subnets for the cluster's Control Plane.
+	// +optional
+	ControlPlaneSubnets []*VPCResourceReference `json:"controlPlaneSubnets,omitempty"`
+
+	// loadBalancers references the VPC Load Balancer's for the cluster.
+	// +optional
+	LoadBalancers []VPCLoadBalancerStatus `json:"loadBalancers,omitempty"`
+
+	// publicGateways references the VPC Public Gateways for the cluster.
+	// +optional
+	PublicGateways []*VPCResourceReference `json:"publicGateways,omitempty"`
+
+	// securityGroups references the VPC Security Groups for the cluster.
+	// +optional
+	SecurityGroups []*VPCResourceReference `json:"securityGroups,omitempty"`
+
+	// vpc references the IBM Cloud VPC.
+	// +optional
+	VPC *VPCResourceReference `json:"vpc,omitempty"`
+}
+
+// VPCResourceReference identifies a resource by id and type.
+type VPCResourceReference struct {
+	// controllerCreated indicates whether the resource is created by the CAPI controller.
+	ControllerCreated bool `json:"controllerCreated,omitempty"`
+
+	// crn defines the IBM Cloud CRN of the resource.
+	CRN string `json"crn,omitempty"`
+
+	// type defines the type of IBM Cloud resource.
+	Type VPCResourceType `json:type"`
+}
+
+type VPCResourceType string
+
+const (
+	VPCResourceTypeCOS               VPCResourceType = VPCResourceType("cos")
+	VPCResourceTypePublicGateway     VPCResourceType = VPCResourceType("publicgateway")
+	VPCResourceTypeSecurityGroup     VPCResourceType = VPCResourceType("securitygroup")
+	VPCResourceTypeSecurityGroupRule VPCResourceType = VPCResourceType("securitygrouprule")
+	VPCResourceTypeSubnet            VPCResourceType = VPCResourceType("subnet")
+	VPCResourceTypeVPC               VPCResourceType = VPCResourceType("vpc")
+)
+```
+
+
+#### Share existing common resource definitions
```

Comment:
@Karthik-K-N this might break few things in PowerVS right, if we take them out of APIs.

IMO, we should focus on only VPC flow and related resources as part of this proposal. Anything related to PowerVS we can deal with later as a separate activity once we are done with VPC flow.